src/db.c

```diff
@@ -1072,7 +1207,16 @@ void scanGenericCommand(client *c, robj *o, unsigned long cursor) {
         };
         do {
             cursor = dictScan(ht, cursor, scanCallback, &data);
-        } while (cursor && maxiterations-- && data.sampled < count);
+            /* In cluster mode there is a separate dictionary for each slot.
+             * If cursor is empty, we should try exploring next non-empty slot. */
+            if (o == NULL && !cursor) {
+                slot = dbGetNextNonEmptySlot(c->db, slot);
+                ht = slot != -1 ? c->db->dict[slot] : NULL;
+            }
+        } while ((cursor || slot > 0) && /* Continue iteration if there are more slots to visit, or cursor hasn't reached the end of dict yet. */
+                 maxiterations-- &&
+                 data.sampled < count);
+        addSlotIdToCursor(slot, &cursor);
```

Comment:
We should probably only encode the slot in the cursor if we're scanning the keyspace, i.e. SCAN, not HSCAN, SSCAN, etc.

```suggestion
        if (o == NULL)
            addSlotIdToCursor(slot, &cursor);
```